File: lib/php/ctrl/mm-action.php
<?php
/*
 * This file is part of the iFilemanager package.
 * (c) 2010-2011 Stotskiy Sergiy <serjo@freaksidea.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
if (!defined('SJ_IS_ADMIN')) {
    header('Location: http://www.google.com');
    exit;
}

$files = isset($_REQUEST['files']) ? $_REQUEST['files'] : array();
if (!is_array($files)) {
    $files = array($files);
}

$rootPath = rtrim($_SERVER['DOCUMENT_ROOT'], DIRECTORY_SEPARATOR);
$fs = new iFilesystem();
$fs->setI18n($_SYSTEM['i18n']);

// %Begin Validate $dirpath
foreach ($files as $filePath => $newName) {
    $pathinfo = $fs->getPathInfo($filePath);
    $dirpath  = ltrim($pathinfo['dirname'], '/');

    $dirpath  = $dirpath ? str_replace('\\', DIRECTORY_SEPARATOR, $dirpath) : '';
    $realpath = realpath($rootPath . DIRECTORY_SEPARATOR . $dirpath);

    $rootLength = strlen($sjConfig['root']);
    $realLength = strlen($realpath);

    if(!$realpath || $realLength < $rootLength){ // hack attemt
        $realpath  = $sjConfig['root'];
        unset($files[$i]);
    }
}
// %End validate $dirpath
require $sjConfig['lib_dir'] . '/model/image.class.php';
$_RESULT['media'] = array();

$width  = $_REQUEST['width'];
$height = $_REQUEST['height'];
$left   = $_REQUEST['left'];
$top    = $_REQUEST['top'];

$overrideOld = !empty($_REQUEST['override']);
$action      = $_REQUEST['action'];

try {
    foreach ($files as $imagePath => $newName) {
        $file  = $rootPath . $imagePath;
        $image = $fs->createImage($file, $overrideOld);
        $pathinfo = $fs->getPathInfo($file);

        switch ($action) {
            case 'resize':
                $image->resize($width, $height);
                break;
            case 'crop':
                $image->resize($width, $height, 'crop', array(
                    'x' => $left,
                    'y' => $top
                ));
                break;
        }

        $image->save($pathinfo['dirname'] . '/' . $newName, 80);
        $_RESULT['media']['add'][] = str_replace($rootPath, '', $image->getSavedName());
        if ($overrideOld) {
            $_RESULT['media']['rm'][]  = $imagePath;
        }
    }
    $_RESULT['response']['status'] = 'success';
} catch (Exception $e) {
	$_RESULT['response']['status'] = 'error';
    $_RESULT['response']['msg']    = $fs->getI18n()->__($e->getMessage());
}
?>

File: lib/ruby/ctrl/mm_action.rb
module SjFileManager
  class MediaController < Controller
    @fs

    def dispatch
      files = get_files

      response = { :media => { :add => [], :rm => [] }, :response => { :status => '', :msg => '' } }

      action = @context.params['action']
      width  = @context.params['width'].to_i
      height = @context.params['height'].to_i
      left   = @context.params['left'].to_i
      top    = @context.params['top'].to_i
      override_old = @context.params['override'] && !@context.params['override'].empty?

      begin
        root = @context.request.env['DOCUMENT_ROOT']
        files.each_pair do |image_path, new_name|
          file  = root + image_path
          pathinfo = fs.get_pathinfo(file)
          img = Magick::Image.read(file).first

          case action
          when 'resize'
            img.resize!(width, height)
          when 'crop'
            img.crop!(left, top, width, height)
          end

          file = pathinfo[:dirname] + File::Separator + fs.prepare_filename(new_name)
          file = fs.dynamic_filename(file) unless override_old

          img.write(file)

          response[:media][:add] << file.sub(root, '')
          response[:media][:rm]  << image_path if override_old
        end
        response[:response][:status] = 'success'
      rescue Magick::ImageMagickError => e
        raise SjException, @context.i18n.__('Unable to modify some images: %s', e.message)
      end

      response
    end

    private
      def fs
        unless @fs
          @fs = Filesystem.new
          @fs.i18n = @context.i18n
        end
        @fs
      end

      def get_files
        files = @context.params['files']

        raise SjException, @context.i18n.__('Unable to process request') unless files.kind_of?(::Hash)

        files.each_pair do |file_path, new_name|
          pathinfo = fs.get_pathinfo(file_path)
          dirpath  = get_dir(pathinfo[:dirname])
          files.delete(file_path) if dirpath.empty?
        end
      end
  end
end